### ISSUE TITLE
FlatDictionary related segfault

### ISSUE THREAD

**Issue by filimonov** [ClickHouse/ClickHouse#9040]
For now i have not more details:

```
2020.02.05 19:18:46.139550 [ 67 ] {} <Fatal> BaseDaemon: ########################################
2020.02.05 19:18:46.139591 [ 67 ] {} <Fatal> BaseDaemon: (version 19.17.4.11 (official build)) (from thread 52) Received signal Segmentation fault (11).
2020.02.05 19:18:46.139610 [ 67 ] {} <Fatal> BaseDaemon: Address: 0x1 Access: read. Address not mapped to object.
2020.02.05 19:18:46.139625 [ 67 ] {} <Fatal> BaseDaemon: Stack trace: 0x5619639a2a20 0x56195fe2b84c 0x56195fe2bdd1 0x56195fe2cdae 0x56195fe2ed71 0x56195fe318dd 0x561961592e6a 0x561961748774 0x561961748bcd 0x56196174c1bf 0x56196174feef 0x56196174d1d7 0x56195e41951c 0x56196419ee00 0x7f82b36fd6ba 0x7f82b302741d
2020.02.05 19:18:46.172333 [ 67 ] {} <Fatal> BaseDaemon: 3. 0x5619639a2a20 memcpy /usr/bin/clickhouse
2020.02.05 19:18:46.172384 [ 67 ] {} <Fatal> BaseDaemon: 4. 0x56195fe2b84c void DB::FlatDictionary::setAttributeValueImpl<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > >(DB::FlatDictionary::Attribute&, unsigned long, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&) /usr/bin/clickhouse
2020.02.05 19:18:46.172401 [ 67 ] {} <Fatal> BaseDaemon: 5. 0x56195fe2bdd1 DB::FlatDictionary::blockToAttributes(DB::Block const&) /usr/bin/clickhouse
2020.02.05 19:18:46.172418 [ 67 ] {} <Fatal> BaseDaemon: 6. 0x56195fe2cdae DB::FlatDictionary::loadData() /usr/bin/clickhouse
2020.02.05 19:18:46.172438 [ 67 ] {} <Fatal> BaseDaemon: 7. 0x56195fe2ed71 DB::FlatDictionary::FlatDictionary(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&, DB::DictionaryStructure const&, std::unique_ptr<DB::IDictionarySource, std::default_delete<DB::IDictionarySource> >, DB::ExternalLoadableLifetime, bool, std::shared_ptr<DB::Block>) /usr/bin/clickhouse
2020.02.05 19:18:46.172450 [ 67 ] {} <Fatal> BaseDaemon: 8. 0x56195fe318dd ? /usr/bin/clickhouse
2020.02.05 19:18:46.172470 [ 67 ] {} <Fatal> BaseDaemon: 9. 0x561961592e6a DB::DictionaryFactory::create(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&, Poco::Util::AbstractConfiguration const&, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&, DB::Context const&) const /usr/bin/clickhouse
2020.02.05 19:18:46.172488 [ 67 ] {} <Fatal> BaseDaemon: 10. 0x561961748774 DB::ExternalDictionariesLoader::create(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&, Poco::Util::AbstractConfiguration const&, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&) const /usr/bin/clickhouse
2020.02.05 19:18:46.172511 [ 67 ] {} <Fatal> BaseDaemon: 11. 0x561961748bcd DB::ExternalLoader::createObject(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&, DB::ExternalLoader::ObjectConfig const&, bool, std::shared_ptr<DB::IExternalLoadable const> const&) const /usr/bin/clickhouse
2020.02.05 19:18:46.172548 [ 67 ] {} <Fatal> BaseDaemon: 12. 0x56196174c1bf std::_Function_handler<std::shared_ptr<DB::IExternalLoadable const> (std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&, DB::ExternalLoader::ObjectConfig const&, bool, std::shared_ptr<DB::IExternalLoadable const> const&), std::_Bind<std::shared_ptr<DB::IExternalLoadable const> (DB::ExternalLoader::*(DB::ExternalLoader*, std::_Placeholder<1>, std::_Placeholder<2>, std::_Placeholder<3>, std::_Placeholder<4>))(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&, DB::ExternalLoader::ObjectConfig const&, bool, std::shared_ptr<DB::IExternalLoadable const> const&) const> >::_M_invoke(std::_Any_data const&, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&, DB::ExternalLoader::ObjectConfig const&, bool&&, std::shared_ptr<DB::IExternalLoadable const> const&) /usr/bin/clickhouse
2020.02.05 19:18:46.172571 [ 67 ] {} <Fatal> BaseDaemon: 13. 0x56196174feef DB::ExternalLoader::LoadingDispatcher::doLoading(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&, unsigned long, bool) /usr/bin/clickhouse
2020.02.05 19:18:46.172596 [ 67 ] {} <Fatal> BaseDaemon: 14. 0x56196174d1d7 ThreadFromGlobalPool::ThreadFromGlobalPool<void (DB::ExternalLoader::LoadingDispatcher::*)(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&, unsigned long, bool), DB::ExternalLoader::LoadingDispatcher*, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&, unsigned long&, bool>(void (DB::ExternalLoader::LoadingDispatcher::*&&)(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&, unsigned long, bool), DB::ExternalLoader::LoadingDispatcher*&&, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&, unsigned long&, bool&&)::{lambda()#1}::operator()() const /usr/bin/clickhouse
2020.02.05 19:18:46.172622 [ 67 ] {} <Fatal> BaseDaemon: 15. 0x56195e41951c ThreadPoolImpl<std::thread>::worker(std::_List_iterator<std::thread>) /usr/bin/clickhouse
2020.02.05 19:18:46.172634 [ 67 ] {} <Fatal> BaseDaemon: 16. 0x56196419ee00 ? /usr/bin/clickhouse
2020.02.05 19:18:46.172649 [ 67 ] {} <Fatal> BaseDaemon: 17. 0x7f82b36fd6ba start_thread /lib/x86_64-linux-gnu/libpthread-2.23.so
2020.02.05 19:18:46.172664 [ 67 ] {} <Fatal> BaseDaemon: 18. 0x7f82b302741d __clone /lib/x86_64-linux-gnu/libc-2.23.so
```

I will add the details if I will get more.

**Comment by filimonov** [comment]
Very general info about use case:
1) dictionary source is ClickHouse
2) dictionary is configured via xml
3) underling table has id field of type UInt32 (not UInt64)
4) crash happen during (probably scheduled by lifetime) reload during the following query:
`INSERT <cols>, dictGetString( ... ) FROM input('....') WHERE ... FORMAT Parquet`

Reproduces not always.

**Comment by alexey-milovidov** [comment]
Similar to #9270

**Comment by filimonov** [comment]
Fix  #10071
